feat: live reloading when running as a module

VishnuSanal opened the pull request:
**Description**

This PR fixes #654

how to run: `python3 -m app --dev`

sansyrox commented:
Hey @VishnuSanal 👋 

I am unable to understand the changes. Could you add an example on why we need the new parameters?

sansyrox commented on robyn/argument_parser.py:
```diff
@@ -26,6 +26,12 @@ def __init__(self) -> None:
             default=False,
             help="Development mode. It restarts the server based on file changes.",
         )
+        parser.add_argument(
+            "--file",
+            dest="file_path",
+            default=None,
+            help="The file associated.",
+        )
```
What does this do?

VishnuSanal replied:
@sansyrox hi, I have changed the implementation. PTAL :)

sansyrox commented:
> Hey @VishnuSanal 👋 
> 
> 
> 
> I am unable to understand the changes. Could you add an example on why we need the new parameters?
> 
> 
> 
> And how you'd invoke the changes?

Hey @VishnuSanal ,
Could you please update the PR description with the above??

VishnuSanal commented:
uh, oh. my previous comment never got sent.

> Could you add an example on why we need the new parameters?

I have changed the implementation, PTAL. removed the need for new params.

> And how you'd invoke the changes?

running as a module usually, like `python3 -m app --dev` would do.

please CMIIW.

sansyrox commented:
@VishnuSanal , I just saw your description. That is not what we want to solve for 😄 

We want to allow the dev mode to work like `robyn -m src --dev`

where src

```
src/
  - __init__.py
  - __main__.py
```

The --dev flag should rerender based on app changes.

VishnuSanal commented:
thank you for the review, Sanskar! this currently works as expected in my testing, PTAL & LMK what you think. :)

sansyrox commented on robyn/argument_parser.py:
```diff
@@ -87,6 +88,8 @@ def __init__(self) -> None:
                 self.file_path = arg
                 break
+        else:
+            self.file_path = sys.argv[0]
```
this is clever. but where are you using this?

VishnuSanal replied:
in `EventHandler#reload` (inside the new `if` block I added)

sansyrox commented on robyn/reloader.py:
```diff
@@ -115,11 +117,19 @@ def reload(self):
-            start_new_session=False,
-        )
+        if self.config.dev and self.config.file_path is not None:
+            subprocess.call(f'cd {"/".join(self.config.file_path.split("/")[-2:-1])}', shell=True)
+            self.process = subprocess.Popen(
+                [sys.executable, "-m", self.config.file_path.split("/")[-2], *arguments],
+                env=new_env,
+                start_new_session=False,
+            )
```
too many things are happening here. Could you split these into different lines?

sansyrox commented on robyn/reloader.py:
```diff
@@ -115,11 +117,19 @@ def reload(self):
+        if self.config.dev and self.config.file_path is not None:
+            subprocess.call(f'cd {"/".join(self.config.file_path.split("/")[-2:-1])}', shell=True)
+            self.process = subprocess.Popen(
+                [sys.executable, "-m", self.config.file_path.split("/")[-2], *arguments],
```
why do we need a `-m` here?

VishnuSanal replied:
to run again as a module whilst reloading...? since, we have run the app as a module initially like `python3 -m src --dev`

sansyrox commented on robyn/argument_parser.py:
```diff
@@ -87,6 +88,8 @@ def __init__(self) -> None:
             if arg.endswith(".py"):
```
@VishnuSanal , this is what stops the app from running in module mode

e.g.

`robyn -m src --dev` will not work if all the code is present in the `src/__main__.py` file

VishnuSanal replied:
> this is what stops the app from running in module mode

yep, the missing parameter is `file_path`. this is why I have assigneed it in the else block & did worksrounds to make it work on other parts of the code.

sansyrox commented on robyn/reloader.py:
```diff
@@ -115,11 +117,24 @@ def reload(self):
-            start_new_session=False,
-        )
+        if self.config.dev and self.config.file_path is not None:
+            split = self.config.file_path.split("/")
+
+            working_directory = "/".join(split[-2:-1])
```
what are we splitting here? could you please write more descriptive names?

sansyrox commented on robyn/reloader.py:
```diff
@@ -115,11 +117,24 @@ def reload(self):
+
+            working_directory = "/".join(split[-2:-1])
+
+            subprocess.call(f"cd {working_directory}", shell=True)
```
you can use `os.chdir` here instead.

Also, why do we need this again?

sansyrox commented on robyn/reloader.py:
```diff
@@ -115,11 +117,24 @@ def reload(self):
+            subprocess.call(f"cd {working_directory}", shell=True)
+
+            self.process = subprocess.Popen(
+                [sys.executable, "-m", split[-2], *arguments],
```
can you assign `split[-2]` to a variable?

sansyrox commented on robyn/__init__.py:
```diff
@@ -10,6 +10,7 @@
 from robyn import status_codes
 from robyn.argument_parser import Config
 from robyn.authentication import AuthenticationHandler
+from robyn.cli import start_dev_server
```
could you move `start_dev_server` to a new file maybe `robyn/dev_server.py`?

sansyrox commented:
Hey @VishnuSanal 👋 

Thanks for the PR, I have some comments 😄

sansyrox commented on robyn/__init__.py:
```diff
@@ -7,10 +7,9 @@
 from nestd import get_all_nested
 
-from robyn import status_codes
+from robyn import status_codes, dev_server
```
@VishnuSanal , please import from `robyn.dev_server` directly here

sansyrox commented on robyn/dev_server.py:
```diff
@@ -0,0 +1,18 @@
+from robyn.argument_parser import Config
+
+
+def start_dev_server(config: Config, file_path: Optional[str] = None):
```
Could you please add a docstring here?

sansyrox commented on robyn/reloader.py:
```diff
@@ -118,14 +118,10 @@ def reload(self):
-            working_directory = "/".join(split[-2:-1])
-
-            subprocess.call(f"cd {working_directory}", shell=True)
+            module_name = self.config.file_path.split("/")[-2]
```
are you sure this will always be the case?

i.e. the module name will always be the second to last name?

VishnuSanal replied:
@sansyrox oh, okay, that makes sense. do you have any suggestions on how to do this?!

here are some solutions from the top of my head:

- can we reload it by running it normally (not a module). but, this is undesirable.

VishnuSanal replied:
hey, my thoughts were, `file_path` contains `sys.argv[0]`. which will be the path to the `__main__.py` file. hence, the second to last of it will be it's directory name. hence the module name. won't this always be the case?

sansyrox replied:
@VishnuSanal , I believe that we can make use of `__file__` that is being passed in the Robyn() class

sansyrox commented on robyn/cli.py:
```diff
@@ -1,11 +1,12 @@
 from InquirerPy.resolver import prompt
 from InquirerPy.base.control import Choice
+
+from . import dev_server
```
please use absolute paths instead of relative paths.

e.g. `from robyn.dev_server import start_dev_server`

sansyrox commented:
Hey @VishnuSanal 👋 

Thanks for the update. I have some follow up comments

sansyrox commented on robyn/cli.py:
```diff
@@ -1,11 +1,12 @@
+
+from robyn.dev_server import start_dev_server
 from .argument_parser import Config
-from .reloader import create_rust_file, setup_reloader
+from .reloader import create_rust_file
```
@VishnuSanal , could we please use absolute paths here as well?

robyn.argument_parser
robyn.reloader

sansyrox commented on robyn/reloader.py:
```diff
@@ -115,11 +117,20 @@ def reload(self):
-            env=new_env,
-            start_new_session=False,
-        )
+        if self.config.dev and self.config.file_path is not None:
```
Can we not use `self.file_path`? 

Then we don't need to pass in `config`.

VishnuSanal replied:
we also need `config.dev`. should I pass it as a variable?

VishnuSanal commented:
current issues:

- doesn't work when using `robyn -m src --dev`. `sys.argv[0]` returns the path to the executable. but, from [here](https://docs.python.org/3/using/cmdline.html#cmdoption-m), it seems like we shouldn't run it like it, but like `python3 -m src --dev`.

(will edit & update this comment...)

VishnuSanal commented:
note to self: 

the flow of control is like: cli#run => cli#start_dev_server => reloader#setup_reloader => EventHandler#reload => Robyn#start and, we are getting the __file__ object at the last stage. my current implementation starts a new dev server from there.

we need a way to bypass it the first time (inorder to reach `robyn#start`) and, then kill the previous reloader & start new one from there.
